### ISSUE TITLE
Handling complex circular objects

### ISSUE THREAD

**Issue by mcollina** [pinojs/pino#1062]
Pino claims to be "safe" and reliable: I would argue one of our principles is that __logging a line should not crash your process__.

Yet we did that and it was fixed in https://github.com/davidmarkclements/fast-safe-stringify/pull/52 and https://github.com/pinojs/pino/pull/1061. The approach of those fixes is to "give up" and avoid serializing such a complex circular object and produce a string instead.

Here is an example:

```js
const s = new PassThrough()
s.resume()
s.write('', () => {})
const obj = { s, p: new Proxy({}, { get () { throw new Error('kaboom') } }) }
const instance = pino()
instance.info({ obj })
```

This is due to the approach that we take in https://github.com/pinojs/pino/blob/14ab3782cb034f82fd7c0013189162c4a0a8c8f3/lib/tools.js#L400-L406: in case of complex objects (like the one above) that cannot be serialized by `JSON.stringify()` due to circular objects, we call `fast-safe-stringify()`.

I would note that this is _not in the hot path_ as the performance of serializing such a complex object is really bad anyway. Moreover, logging a line should not crash a process. The reason for the problem was that `fast-safe-stringify` replaces the keys with circular references.  It's a pretty risky approach to take in something that is not a hot path and needed for the developer experience.

One approach to avoid this kind of problems in the future is to use https://github.com/BridgeAR/safe-stable-stringify from @BridgeAR instead of fss. We could potentially look into some other safe-but-slow algorithms.

**Comment by mcollina** [comment]
WDYT @jsumners @davidmarkclements ?

**Comment by jsumners** [comment]
Ship it.

**Comment by davidmarkclements** [comment]
this makes sense to me

**Comment by davidmarkclements** [comment]
closed by #1066
